### PR TITLE
Disconnect Peers based on ForkId validations

### PR DESCRIPTION
**Pull request by smartprogrammer93**
## Changes

Add validation to peers based on ForkId. Most work here is done by @asdacap . I only reviewed and added a new way of validation.

#### What types of changes does your code introduce?

- [ ] Bugfix (a non-breaking change that fixes an issue)
- [x] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a fix or a feature that would cause existing functionality not to work as expected)
- [ ] Documentation update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional or API changes)
- [ ] Build-related changes
- [ ] Other: _description_ 

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [x] Yes
- [ ] No

## Remarks
I have added a new approach that i think should be considered instead of the Geth approach

### CODE REVIEW

File: src/Nethermind/Nethermind.Network/ForkInfo.cs [hunk 1]
```diff
@@ -70,5 +76,151 @@ public ForkId GetForkId(long headNumber, ulong headTimestamp)
...
+            //        information.
+            //   4. Reject in all other cases.
+            BlockHeader? head = _blockTree.Head?.Header;
+            if (head == null) return IForkInfo.ValidationResult.Valid;
+
+            for (int i = 0; i < Forks.Length; i++)
+            {
+                (ForkActivation forkActivation, ForkId forkId) = Forks[i];
+                bool usingTimestamp = forkActivation.Timestamp != null;
+                ulong headActivation = (usingTimestamp ? head.Timestamp : (ulong)head.Number);
```
Comment (asdacap) [inline src/Nethermind/Nethermind.Network/ForkInfo.cs:113]: Did you double check this with geth's edge case? https://github.com/ethereum/go-ethereum/pull/25878/files#diff-77f77794b42f4fd84db10f281103146d1f12e9b50a404254bc87565e3e423872R193. 

Did you also copy the new test case? Coz I did not.

Comment (smartprogrammer93) [reply src/Nethermind/Nethermind.Network/ForkInfo.cs:117]: Now it is. check it out @asdacap

**Comment by tkstanczak** [comment]
it was not used in the past since Nethermind was smart in syncing archive nodes ancient blocks from ETC clients that were more readily serving them than Geths
here you disconnect them forever but maybe that functionality got lost anyway

**Comment by tkstanczak** [comment]
I was never a fan of disconnecting based on ForkId - so I just implemented announcing it but not disconnecting

**Comment by smartprogrammer93** [comment]
@tkstanczak we are only implementing it because it is being included in the spec. So before disconnecting was the client's choice to implement on not, but i think once it's in spec we should probably implement it. we can always add a flag that can disable it in the config if this is something you feel strongly about.

**Comment by LukaszRozmej** [comment]
Disconnecting on ForkId is very useful on actual forks and even more on shadow forks to actually have the relevant peers in the list. It is also in spec, which we want to follow more rigorously.

File: src/Nethermind/Nethermind.Network/ForkInfo.cs [hunk 2]
```diff
@@ -3,36 +3,51 @@
...
+            DictForks.Add(BitConverter.ToUInt32(hash), toAdd);
             for (int index = 0; index < transitionActivations.Length; index++)
             {
                 ForkActivation forkActivation = transitionActivations[index];
                 BinaryPrimitives.WriteUInt64BigEndian(blockNumberBytes, forkActivation.Activation);
                 hash = CalculateHash(ref crc, blockNumberBytes);
-                Forks[index + 1] = (forkActivation, new ForkId(hash, GetNextActivation(index, transitionActivations)));
+                toAdd = (forkActivation, new ForkId(hash, GetNextActivation(index, transitionActivations)));
+                Forks[index + 1] = toAdd;
+                DictForks.Add(BitConverter.ToUInt32(hash), toAdd);
```
[the remark this answers is not in the document]
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Network/ForkInfo.cs:50]: Btw can we just use `uint` (`crc`) as our `hash` ?

File: src/Nethermind/Nethermind.Network/ForkInfo.cs [hunk 3]
```diff
@@ -70,5 +85,142 @@ public ForkId GetForkId(long headNumber, ulong headTimestamp)
...
+
+            throw new InvalidOperationException();
+        }
+
+        /// <summary>
+        /// Verify that the forkid from peer matches our forks. Code is largely copied from Geth.
+        /// </summary>
+        /// <param name="peerId"></param>
+        /// <returns></returns>
+        public IForkInfo.ValidationResult ValidateForkId2(ForkId peerId)
```
[the remark this answers is not in the document]
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Network/ForkInfo.cs:183]: I like 2 a LOT more, avoids linear scan, basically what method 1 does is:
1. Find local forkId based on head with linear scan
2. Compare it with remote forkId, potentially return Valid/IncompatibleOrStale
3. Try find our past fork that is based on peer forkId with linear scan, potentially return Valid/RemoteStale
4. Try find our future fork that is based on peer forkId with linear scan, return Valid/IncompatibleOrStale

Version 2 skips the linear scan, going straight to the matching fork, then checks:
1. If next forks don't match:
2. if this is last remotePeer fork and for us it is not and we activated next one - RemoteStale
3. if peer has next fork and we also activated our fork:
- if we would also activate peerFork -> IncompatibleOrStale
- if we have next fork and we would activate it -> IncompatibleOrStale

This would benefit from better naming, maybe handling all options explicitly?

File: src/Nethermind/Nethermind.Network/ForkInfo.cs [hunk 4]
```diff
@@ -70,5 +85,142 @@ public ForkId GetForkId(long headNumber, ulong headTimestamp)
...
+                && (peerId.Next == 0
+                || peerId.Next >= TimstampThreshold);
+            ulong headActivation = (usingTimestamp ? head.Timestamp : (ulong)head.Number);
+
+            // my approach is to accept all peers except the ones we dont like. which is the oposite of what
+            // geth does. they reject all except ones they think are right.
+
+            if (found.Id.Next != peerId.Next)
+            {
+                if (peerId.Next == 0
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network/ForkInfo.cs:202]: might be easier to read when named correctly, suggestions (not sure if 100% correct):
`peerId.Next == 0` -> `peerOnLastFork`
`found.Id.Next == 0` -> `localOnLastFork`
`found.Id.Next > 0` -> `!localOnLastFork`
`headActivation >= found.Id.Next` -> `headPastLastFork`

File: src/Nethermind/Nethermind.Network/ForkInfo.cs [hunk 5]
```diff
@@ -70,5 +85,142 @@ public ForkId GetForkId(long headNumber, ulong headTimestamp)
...
+                if (peerId.Next > 0
+                    && headActivation > found.Activation.Activation)
+                {
+                    if (headActivation >= peerId.Next)
+                    {
+                        return IForkInfo.ValidationResult.IncompatibleOrStale;
+                    }
+                    if (found.Id.Next > 0
+                        && headActivation >= found.Id.Next
+                        && peerId.Next != found.Id.Next)
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network/ForkInfo.cs:217]: last && part is the same as most outer if - so it is redundant?

File: src/Nethermind/Nethermind.Network/ForkInfo.cs [hunk 6]
```diff
@@ -3,36 +3,51 @@
...
 {
-    public class ForkInfo
+    public class ForkInfo : IForkInfo
     {
+        private Dictionary<uint, (ForkActivation Activation, ForkId Id)> DictForks { get; }
         private (ForkActivation Activation, ForkId Id)[] Forks { get; }
 
-        public ForkInfo(ISpecProvider specProvider, Keccak genesisHash)
+        private IBlockTree _blockTree;
+        private ulong TimstampThreshold => 1438269973ul; // MainnetSpecProvider.ShanghaiBlockTimestamp;
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Network/ForkInfo.cs:27]: We can't just bake it here! What about other networks like Gnosis?

Comment (smartprogrammer93) [reply src/Nethermind/Nethermind.Network/ForkInfo.cs:27]: Why not, it is not like there any other networks that have timestamp forks before this

File: src/Nethermind/Nethermind.Network/ProtocolValidator.cs [hunk 7]
```diff
@@ -20,58 +20,113 @@ public class ProtocolValidator : IProtocolValidator
...
+            {
+                if (_logger.IsTrace)
+                    _logger.Trace(
+                        $"Initiating disconnect with peer: {session.RemoteNodeId}, different genesis hash: {syncPeerArgs.GenesisHash}, our: {_blockTree.Genesis.Hash}");
+                _nodeStatsManager.ReportFailedValidation(session.Node, CompatibilityValidationType.DifferentGenesis);
+                Disconnect(session, InitiateDisconnectReason.InvalidGenesis, "invalid genesis");
+                if (session.Node.IsStatic && _logger.IsWarn)
+                    _logger.Warn(
+                        $"Disconnected an invalid static node: {session.Node.Host}:{session.Node.Port}, reason: {InitiateDisconnectReason.InvalidGenesis} (invalid genesis)");
+                return false;
```
[the remark this answers is not in the document]
Comment (smartprogrammer93) [reply src/Nethermind/Nethermind.Network/ProtocolValidator.cs:90]: i tried, there is just too many parameters

File: src/Nethermind/Nethermind.Network/ForkInfo.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (smartprogrammer93) [reply src/Nethermind/Nethermind.Network/ForkInfo.cs]: using uint now

**Review by LukaszRozmej — APPROVED** [review]

**Review by LukaszRozmej — APPROVED** [review]
